docs: Querying for images with GraphQL and gatsby-image (#16566)
gatsbyjs/gatsby: docs/docs/recipes.md

- [Using the Static Folder](/docs/static-folder/)
- [More on all image techniques in Gatsby](/docs/images-and-files/)

### Optimizing and querying local images with gatsby-image

The `gatsby-image` plugin can take out much of the pain of optimizing images for your site. Once installed, Gatsby will generate optimized resources to improve loading. Those resources can be queried with GraphQL and passed into Gatsby's image component to do the heavy lifting of creating different image sizes and loading them at the right times for you.

#### Prerequisites

- the `gatsby-image`, `gatsby-transformer-sharp`, and `gatsby-plugin-sharp` packages installed
- [images sourced](/packages/gatsby-image/#install) in your `gatsby-config` with a plugin like `gatsby-source-filesystem`

#### Directions

> 1. First, import `Img` from `gatsby-image`, as well as `graphql` and `useStaticQuery` from `gatsby` into a component

Reviewer: ```suggestion
1. First, import `Img` from `gatsby-image`, as well as `graphql` and `useStaticQuery` from `gatsby`
```

```jsx
import { useStaticQuery, graphql } from "gatsby" // to query for image data
import Img from "gatsby-image" // to take image data and render it
```

> 2. Write a query to get image(s) data, and provide the data to the `<Img />` component:

Reviewer: ```suggestion
2. Write a query to get image data, and pass the data into the `<Img />` component:
```

Choose any of the following options or a combination of them.

> a. for a single image based on its path from where the images are [sourced](/docs/content-and-data/) (ex: `images/corgi.jpg`)

Reviewer: ```suggestion
a. a single image based queried by its file [path](/docs/content-and-data/) (Example: `images/corgi.jpg`)
```

```jsx
const data = useStaticQuery(graphql`
  query {
    file(relativePath: { eq: "corgi.jpg" }) { // highlight-line
      childImageSharp {
        fluid {
          base64
          aspectRatio
          src
          srcSet
          sizes
        }
      }
    }
  }
`)

return (
  <Img fluid={data.file.childImageSharp.fluid} alt="A corgi smiling happily" />
)
```

> b. with a [GraphQL fragment](/docs/using-fragments/), querying for the necessary fields more tersely

Reviewer: ```suggestion
b. using a [GraphQL fragment](/docs/using-fragments/), to query for the necessary fields more tersely
```

```jsx
const data = useStaticQuery(graphql`
  query {
    file(relativePath: { eq: "corgi.jpg" }) {
      childImageSharp {
        fluid {
          ...GatsbyImageSharpFluid // highlight-line
        }
      }
    }
  }
`)

return (
  <Img fluid={data.file.childImageSharp.fluid} alt="A corgi smiling happily" />
)
```

> c. for a directory (ex. `images/dogs`) of images by [filtering](/docs/graphql-reference/#filter) for the extension and relative directory fields, and then mapping the results into `Img` components

Reviewer: ```suggestion
c. several images from a directory (Example: `images/dogs`) [filtered](/docs/graphql-reference/#filter) by the `extension` and `relativeDirectory` fields, and then mapped into `Img` components
```

```jsx
const data = useStaticQuery(graphql`
  query {
    allFile(
      // highlight-start
      filter: {
        extension: { regex: "/(jpg)|(png)|(jpeg)/" }
        relativeDirectory: { eq: "dogs" }
      }
      // highlight-end
    ) {
      edges {
        node {
          base
          childImageSharp {
            fluid {
              ...GatsbyImageSharpFluid
            }
          }
        }
      }
    }
  }
`)

return (
  <div>
    // highlight-start
    {data.allFile.edges.map(image => (
      <Img
        fluid={image.node.childImageSharp.fluid}
        alt={image.node.base.split(".")[0]} // only use section of the file extension with the filename
      />
    ))}
    // highlight-end
  </div>
)
```

> **Note**: with this method it is difficult to match images to `alt` text for accessibility. This example uses images with alt text encoded in the filenames like `dog in a party hat.jpg`.

Reviewer: ```suggestion
**Note**: This method can make it difficult to match images with `alt` text for accessibility. This example uses images with `alt` text included in the filename, like `dog in a party hat.jpg`.
```

> d. for an image of a fixed size with the `fixed` field, instead of `fluid`

Reviewer: ```suggestion
d. an image of a fixed size using the `fixed` field instead of `fluid`
```

```jsx
const data = useStaticQuery(graphql`
  query {
    file(relativePath: { eq: "corgi.jpg" }) {
      childImageSharp {
        fixed(width: 250, height: 250) { // highlight-line
          ...GatsbyImageSharpFixed
        }
      }
    }
  }
`)
return (
  <Img fixed={data.file.childImageSharp.fixed} alt="A corgi smiling happily" />
)
```

> e. for an image of a fixed size with `maxWidth`

Reviewer: ```suggestion
e. an image of a fixed size with a `maxWidth`
```

```jsx
const data = useStaticQuery(graphql`
  query {
    file(relativePath: { eq: "corgi.jpg" }) {
      childImageSharp {
        fixed(maxWidth: 250) { // highlight-line
          ...GatsbyImageSharpFixed
        }
      }
    }
  }
`)
return (
  <Img fixed={data.file.childImageSharp.fixed} alt="A corgi smiling happily" /> // highlight-line
)
```

> f. for an image filling a fluid container with a max width (800px) and a higher quality (75%, the default value is 50 which equates to 50%)

Reviewer: ```suggestion
f. an image filling a fluid container with a max width (in pixels) and a higher quality (the default value is 50 i.e. 50%)
```

```jsx
const data = useStaticQuery(graphql`
  query {
    file(relativePath: { eq: "corgi.jpg" }) {
      childImageSharp {
        fluid(maxWidth: 800, quality: 75) { // highlight-line
          ...GatsbyImageSharpFluid
        }
      }
    }
  }
`)

return (
  <Img fluid={data.file.childImageSharp.fluid} alt="A corgi smiling happily" />
)
```

> 3. (Optional) you can add inline styles to the `<Img />` like you would to other elements

Reviewer: ```suggestion
3. (Optional) Add inline styles to the `<Img />` like you would to other components
```

```jsx
<Img
  fluid={data.file.childImageSharp.fluid}
  alt="A corgi smiling happily"
  style={{ border: "2px solid rebeccapurple", borderRadius: 5, height: 250 }} // highlight-line
/>
```

> 4. (Optional) you can force an image into a desired aspect ratio by overriding the `aspectRatio` field returned from the GraphQL query that gets passed into the `<Img />` component

Reviewer: ```suggestion
4. (Optional) Force an image into a desired aspect ratio by overriding the `aspectRatio` field returned by the GraphQL query before it is passed into the `<Img />` component
```

```jsx
<Img
  fluid={{
    ...data.file.childImageSharp.fluid,
    aspectRatio: 1.6, // 1280 / 800 = 1.6
  }}
  alt="A corgi smiling happily"
/>
```

> 5. Run `gatsby develop`, which will generate images for files sourced in the filesystem (if it hasn't already done so, and cached them)

Reviewer: ```suggestion
5. Run `gatsby develop`, to generate images from files in the filesystem (if not done already) and cache them
```

#### Additional Resources

> - [Example repo implementing these examples](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipes-gatsby-image)

Reviewer: ```suggestion
- [Example repository illustrating these examples](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipes-gatsby-image)
```

- [Gatsby Image API](/docs/gatsby-image/)
- [Using Gatsby Image](/docs/using-gatsby-image)
- [More on working with images in Gatsby](/docs/working-with-images/)

### Optimizing and querying images in post frontmatter with gatsby-image

> For use cases like a featured image for a blog post, you can still use `gatsby-image`. The `gatsby-image` component needs to be given processed image data, which can come from a file locally or remote, including from a url in the frontmatter of a `.md` or `.mdx` file.

Reviewer: ```suggestion
For use cases like a featured image in a blog post, you can _still_ use `gatsby-image`. The `Img` component needs processed image data, which can come from a local (or remote) file, including from a URL in the frontmatter of a `.md` or `.mdx` file.
```

> For inline images inside of markdown (using the `![]()` markdown syntax), consider a plugin like [`gatsby-remark-images`](/packages/gatsby-remark-images/)

Reviewer: ```suggestion
To inline images in markdown (using the `![]()` syntax), consider using a plugin like [`gatsby-remark-images`](/packages/gatsby-remark-images/)
```

#### Prerequisites

> - the `gatsby-image`, `gatsby-transformer-sharp`, and `gatsby-plugin-sharp` packages installed

Reviewer: ```suggestion
- The `gatsby-image`, `gatsby-transformer-sharp`, and `gatsby-plugin-sharp` packages installed and added to the plugins array in `gatsby-config`
```

> - [images sourced](/packages/gatsby-image/#install) in your `gatsby-config` with a plugin like `gatsby-source-filesystem`

Reviewer: ```suggestion
- [Images sourced](/packages/gatsby-image/#install) in your `gatsby-config` using a plugin like `gatsby-source-filesystem`
```

> - markdown files sourced in your `gatsby-config` with an image url in the post frontmatter

Reviewer: ```suggestion
- Markdown files sourced in your `gatsby-config` with image URLs in frontmatter
```

> - [pages created](/docs/creating-and-modifying-pages/) from sourced markdown with the `createPages` method

Reviewer: ```suggestion
- [Pages created](/docs/creating-and-modifying-pages/) from Markdown using [`createPages`](https://www.gatsbyjs.org/docs/node-apis/#createPages)
```

#### Directions

> 1. Verify that your markdown file has an image url with a correct path to an image in your project

Reviewer: ```suggestion
1. Verify that the Markdown file has an image URL with a valid path to an image file in your project
```

```mdx:title=post.mdx
---
title: My First Post
featuredImage: ./corgi.png // highlight-line
---

Post content...
```

> 2. Verify that a unique identifier (like a slug in this case) is passed into the context of the call to `createPages` in your `gatsby-node.js`, which will be passed into the GraphQL query in the layout component as a variable `slug`

Reviewer: ```suggestion
2. Verify that a unique identifier (a slug in this example) is passed in context when `createPages` is called in `gatsby-node.js`, which will later be passed into a GraphQL query in the Layout component
```

```js:title=gatsby-node.js
exports.createPages = async ({ graphql, actions }) => {
  const { createPage } = actions

  // query for all markdown

  result.data.allMdx.edges.forEach(({ node }) => {
    createPage({
      path: node.fields.slug,
      component: path.resolve(`./src/components/markdown-layout.js`),
      // highlight-start
      context: {
        slug: node.fields.slug,
      },
      // highlight-end
    })
  })
}
```

> 3. Then, import `Img` from `gatsby-image`, and `graphql` from `gatsby` into the template component being used for your markdown, write a [pageQuery](/docs/page-query/) to get image data based off of the `slug` passed in through the page context, and provide the data to the `<Img />` component:

Reviewer: ```suggestion
3. Now, import `Img` from `gatsby-image`, and `graphql` from `gatsby` into the template component, write a [pageQuery](/docs/page-query/) to get image data based on the passed in `slug` and pass that data to the `<Img />` component:
```

```jsx:title=markdown-layout.jsx
import React from "react"
import { graphql } from "gatsby" // highlight-line
import Img from "gatsby-image" // highlight-line

export default ({ children, data }) => (
  <main>
    // highlight-start
    <Img
      fluid={data.markdown.frontmatter.image.childImageSharp.fluid}
      alt="A corgi smiling happily"
    />
    // highlight-end
    {children}
  </main>
)

// highlight-start
export const pageQuery = graphql`
  query PostQuery($slug: String) {
    markdown: mdx(fields: { slug: { eq: $slug } }) {
      id
      frontmatter {
        image {
          childImageSharp {
            fluid {
              ...GatsbyImageSharpFluid
            }
          }
        }
      }
    }
  }
`
// highlight-end
```

4. Run `gatsby develop`, which will generate images for files sourced in the filesystem

#### Additional Resources

> - [Example repo implementing this recipe](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipes-gatsby-image)

Reviewer: ```suggestion
- [Example repository using this recipe](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipes-gatsby-image)
```
